docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,532 @@
+
+    > AWS APIs have different behavior for returning the late initialized fields to the user. Some services return these fields as part of Create output, while some services will return the late initialized fields as part of  ReadOne output after the resource creation.
+
+3. Wait for asynchronous resource creation to populate late initialized fields.
+
+    > For some services/resources the late initialized fields do not appear immediately.  This happens when there is asynchronous processing involved in resource creation. The proposed solution should allow waiting and retry until the late initialized fields appear in the mentioned API output.
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Proposed Solution
+In this solution, There will be updates in 
+1) **ACK Code Generator** to capture the late initialized fields from AWS operation output and use them to late initialize k8s custom resource object. Code generator changes will support a) configuration for service teams to tell ACK code generator which fields will be late initialized and b) configuration to provide custom code for late initialization.
+2) **ACK Runtime** to a) update the custom resource stored in etcd with late initalized optional spec fields And b) handle the late initalized fields for resources that have asynchronous creation.
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. Fix the latent bug where Spec fields are not set from the Create and Update response. Once this bug is fixed the late initialized fields which are present in Create and Update response will be set.
+    ```go
+    // Current Code
+    // sdk.go.tpl    
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 false }}
+        ...
+    }
+    ```
+
+    ```go
+    // Proposed Code
+    // sdk.go.tpl
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        // NOTE: The performSpecUpdate boolean param will be removed
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 }}
+        ...
+    }
+    ```
+
+2. **Provide a new "LateInitializationConfig" struct inside generator config for AWS service teams to provide which spec fields will be late initialized.** These fields will be the ones which are late initialized by the server but not present in the Create and Update response.
+
+    These fields will be populated by making the readOne call. AWS service teams can also provide a delay after which these fields will be late initialized.
```

Comment:
Your requirements outline that there should be sufficient wait/retry, but this seems to only specify a wait. How many attempts to read the late initialised fields will the reconciler make? Is this configurable?

As I understand for some resources there may be a very long delay for some fields - while waiting for the underlying instance to boot, for instance. In those cases, rather than specifying a 15 minute delay, I'd prefer to have it delay and retry every, say, 2 minutes - finally failing after a 15 minute total.

Reply:
IMO and at least in the first implementation, we should keep it as simple as possible as we don't have any similar behavior in any of the controllers.